Allow redeploying to interactive tests

Radvendii opened the pull request:
## Description of changes

- [x] adds a `rebuild()` command to the NixOS testing driver, that will rebuild the test and push any changes to running VMs
- [x] adds a `--rebuild-cmd` flag to the driver to specify what command rebuilds the test (can also be passed directly to `rebuild()`
- [x] adds a `--internal-print-update-driver-info-and-exit` flag to the driver that will just print the information needed to update another driver.
- [x] document the changes
- [x] do more extensive stress-testing
  - [x] adding machines
  - [x] removing machines
  - [x] multi-machine configs
    - [x] adding one machine and removing another
    - [x] removing a machine changes the other machines' configs accordingly
  - [x] changing vlans
  - [x] changing `testScript`

Bugs
- [x] The vlans are getting updated on every `rebuild()` for some reason

This PR has been reworked since it's original formulation. See below for the original text.

<details>
<summary>Original PR description</summary>

- [x] adds a command to the output of `driverInteractive`, as well as in its own output `redeploy`, which will deploy the current test machine configuration to an actively running `driverInteractive`
- [x] Adds an extra machine to `driverInteractive` which serves as a jump host so that the user can ssh into machines in the test (including to do the redeploy).
- [x] Adds an ssh config to the output of `driverInteractive`, as well as in its own output `sshConfig` to take advantage of that jump host. This can be used with `ssh -F ./result/ssh_config`
- [ ] document changes (haven't yet done. waiting for feedback)
- [ ] automatically start jump host. Currently this setup requires the manual step of running `redeploy_jumphost.start()` after starting the interactive driver (couldn't figure out how to do this. does anyone have ideas?)
</details>

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Radvendii commented:
I opened the issue (#281332) for discussion of alternate solutions and whether this is an issue worth solving at all.

Radvendii commented:
Tagging some people who have touched the testing infrastructure recently

@roberth @K900 @tfc

roberth commented on nixos/lib/testing/interactive.nix:
```diff
@@ -1,6 +1,97 @@
+        exit 1
+      fi
+
+      # taken from nixos-rebuild (we only want to do the activate part)
+       cmd=(
+          "systemd-run"
+          "-E" "LOCALE_ARCHIVE"
+          "--collect"
+          "--no-ask-password"
+          "--pty"
+          "--quiet"
+          "--same-dir"
+          "--service-type=exec"
+          "--unit=nixos-rebuild-switch-to-configuration"
+          "--wait"
+          "''${configPaths[$machine]}/bin/switch-to-configuration"
+          "test"
+      )
```
- Why `test` instead of `switch`? Shouldn't a reboot of the VM boot the new config?
- This `systemd-run` invocation should not be duplicated, but be factored out. I've proposed `${toplevel}/bin/apply` in https://github.com/NixOS/nixpkgs/issues/266290

Radvendii replied:
- `test` rather than `switch` because the VMs don't have bootloaders, so `switch` will fail.
- are you proposing we hold off on this until #266290 is merged, or that I should work on it myself, or that there's some other way we should factor it out in the meantime in this PR? I'm happy with any of them (except holding off indefinitely if it's not likely to be merged soon)

roberth commented on nixos/lib/testing/interactive.nix:
```diff
@@ -1,6 +1,97 @@
+      ProxyJump redeploy_jumphost
+  '';
+
+  redeploy = hostPkgs.writeShellScriptBin "redeploy" ''
```
I *think* I'd prefer for this to be implemented in the (python) driver instead, but I'll defer to @tfc.

Radvendii replied:
I'd love if `redeploy` could be put in the python driver. It could maybe be more powerful in that case, since it could also deploy new machines / power off ones that have been decommissioned.

But I'm not sure how that's possible. The python driver only has access to the configuration that existed at the time it was spun up. The whole point is that we then reconfigure the machines, and while the old driver is still running, redeploy the new configurations.

roberth commented:
Question: The `nodes` are observable by the rest of the test, which will probably lead to heisenbugs. Assuming there isn't a command that just forwards socket connections to an address and port on a VDE network, perhaps you could use a different option for this extra node config; not `nodes.<something>`?

tfc commented:
It feels like this should neither increase the complexity of the python code nor of the standard test module code.
Can we maybe have this similar to `profiles` in nixpkgs? So that it is not included by default and people who don't use it don't have to regard it?

The reason for my thinking is that this is quite an opinionated way to do it (i.e. not switching on a few flags but even adding another host that ought to be used as jumphost etc.) and might not automatically play well with any test.

Radvendii commented:
Points well taken. To summarize

1. This is a disruptive way of doing this, such that it shouldn't be on by default.

2. Mostly this is because it adds an extra machine

I will work on finding a way around (2). But if there isn't a clean way to do it in the end, maybe the solution is more to put this in NUR (or a flake), and in Nixpkgs the only change necessary would be to expose the underlying module system structure of the tests, so that it can be extended outside of Nixpkgs.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-55/40996/1

RaitoBezarius commented:
cc @nikstur as folks working on the NixOS tests ecosystem have not been paged.

K900 commented:
Hard disagree. If we have slow iteration times, we should be removing complexity, not adding more complexity to bypass original complexity.

Radvendii commented:
It's not complexity that leads to the slow iteration time, it's the time to restart the VM and spin up all the services again. For some tests this is quick, but for others it can be slow and there's not much we can do to change that if the test needs services that are slow to start.

There's also the issue of: if I am running the interactive driver, and making stateful changes to test things out, and then I want to make a change to the configuration and test that out, I now have to redo all of the stateful changes I made. It's a question of workflow getting disrupted.

Radvendii commented:
There's another approach demonstrated here: https://github.com/tweag/nix-hour/blob/master/templates/vm/configuration.nix#L164-L167

Which mounts the relevant directory into the VM so that you can run some version of `nixos-rebuild` from *inside* the VM and it has the latest version of the source code.

This only works though because the directory structure is highly constrained. NixOS tests, and really any NixOS VM  are Nix derivations that can come from, in theory, any Nix code, so you can't know how to extract the derivation from the directory, or even which directory to mount. There could even be references to arbitrary other parts of the file system.

I don't see a way to solve that problem...

Radvendii committed:
update from the driver

Radvendii committed:
add --rebuild-cmd flag

Radvendii committed:
document

Radvendii commented:
Okay, I've completley reworked how this PR functions. No jumphosts, no sshing, no modification of the machines in the test.

Instead, I've added a method to the driver, `rebuild()` which will run a rebuild command and then get information from the produced executable about how to update the driver.

It then does that update, by changing fields on the machines, adding new machines and removing old ones, and running nixos-rebuild on machines whose configurations have changed.

I've also added a flag to the driver for internal use which just produces that information needed to update.

There's a little bit of "magic" going on where we assume the path of the new driver executable will be the same as the one we're running now. This works well with `nix-build` and `nix build` which modify the `result/` symlink, but for instance it will not work if you `nix run .#nixosTests.foo.driverInteractive` directly, since no `result/` symlink is produced. I think this is an okay balance of convenience to magic, especially since you can override the `exe` path if needed.

Radvendii committed:
fix vlans issue

tfc commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -101,6 +106,122 @@ def __exit__(self, *_: Any) -> None:
+                if machine.name not in names:
+                    if machine.is_up():
+                        self.logger.warning(
+                            f"{machine.name} removed from the test, but it's running so we're not going to shut it down automatically. Call driver.machines[\"{machine.name}\"].stop() and re-run rebuild() to remove."
```
If I understand correctly, with N machines running, we still need to stop all of them before executing `rebuild(...)`. This approach doesn’t seem significantly more advantageous than stopping the entire driver and rerunning it with the `--keep-vm-state` parameter. This way, the interactive shell history and the VMs’ disks remain intact for the next run.

Could you please elaborate on your specific use case and the exact benefits you see from this additional complexity? Understanding this will help evaluate if the added complexity is justified.

Radvendii replied:
> If I understand correctly, with N machines running, we still need to stop all of them before executing rebuild(...)

No! This is only if you **delete** a machine from the test, communicating that you no longer want it to be there at all.

If you **change** the config of a machine, it will do the equivalent of running `nixos-rebuild switch` on that machine. Namely, it will build the new configuration and run `<new-configuration-store-path>/system/bin/switch-to-configuration test` on the running VM

Radvendii replied:
I expect deleting machines from the test in this context will be quite uncommon. This code is just handling an edge case gracefully. Most of the time you will be modifying existing machines, and that's where this PR shines, since it allows you to redeploy the VMs in-place rather than restart them.

roberth commented on nixos/doc/manual/development/running-nixos-tests-interactively.section.md:
```diff
@@ -24,6 +24,23 @@ back into the test driver command line upon its completion. This allows
 
+## Separating outputs {#sec-nixos-test-separating-outputs}
+
+Logs from the running virtual machines get output in-line and can be disruptive
+while using the python REPL. You can separate these streams by redirecting
+stderr:
```
The documentation team recommends one sentence per line, which can be a long line. It works better with GitHub suggestions, as it requires no reflowing of these ignored line breaks.
(Pre-existing problem and not urgent for these additions)

roberth commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -101,6 +106,122 @@ def __exit__(self, *_: Any) -> None:
+                                Path(start_cmd._cmd).parent.parent
+                                / "system"
+                                / "bin"
+                                / "switch-to-configuration"
```
The final part of this path is documented in a couple of places in the NixOS manual, so that won't be much of a problem.

roberth commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -101,6 +106,122 @@ def __exit__(self, *_: Any) -> None:
+                            machine.start_command = start_cmd
+                else:
+                    self.logger.warning(
+                        f"{start_cmd.machine_name} has multiple instances. This shouldn't be possible, but either way it's now ambiguous which to update."
```
Could we log info about the running instances?

Radvendii replied:
I'm not sure how this would happen so I'm not sure what would be useful info. If it does happen it's certainly a bug in the testing framework and not a problem with the user's code. Maybe I should say that explicitly.

roberth commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -101,6 +106,122 @@ def __exit__(self, *_: Any) -> None:
+                            machine.start_command = start_cmd
+                else:
+                    self.logger.warning(
+                        f"{start_cmd.machine_name} has multiple instances. This shouldn't be possible, but either way it's now ambiguous which to update."
```
```suggestion
                        f"Skipping the update of {start_cmd.machine_name}, because it has multiple instances. This shouldn't be possible, but either way it's now ambiguous which to update."
```

roberth commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -101,6 +106,122 @@ def __exit__(self, *_: Any) -> None:
+                if nr not in old_nrs:
+                    self.vlans.append(VLan(nr, tmp_dir, self.logger))
+
+        # if the test script has already been run we can't post-hoc modify the actions, but we can at least update the script
```
Maybe log this?

roberth replied:
Does this mean `test_script()` in the repl will launch the new script?

Radvendii replied:
Yep, that's the idea.

Radvendii replied:
Updated the log message and removed the comment.

roberth commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -101,6 +106,122 @@ def __exit__(self, *_: Any) -> None:
+                                / "bin"
+                                / "switch-to-configuration"
+                            )
+                            machine.succeed(f"{switch_cmd} test")
```
This only works if the host store is forwarded into the VM.
If the VM is completely image based, the new store paths will have to be copied in.
What if we always forward the host store to `/root/host-store` in the guests? No sane program will find it there, so it doesn't affect the purity or performance of the test, but it does allow us to do a `nix copy` just for the purpose of making this work.
This should error out early if https://github.com/NixOS/nixpkgs/blob/883180e6550c1723395a3a342f830bfc5c371f6b/nixos/modules/virtualisation/qemu-vm.nix#L771

Radvendii replied:
We don't have access to the nixos configuration here, only the built output. We could check for the presence of the store path, and if it doesn't exist produce a helpful error message that points to this as a likely culprit. Were you imagining something else?

Always forwarding the host store to `/root/host-store` seems like a big change for only this purpose. It seems not very common to set `useNixStoreImage` in tests, and IIUC it's only for performance purposes, so it can be turned off while interactively developing the test.

That being said, my feelings against adding `/root/host-store` are not strong, so if you or other people feel like it is a good change, I'm happy to add it in.

Radvendii committed:
Apply suggestions from code review

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

Radvendii committed:
more updates from review

YorikSar commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -101,6 +106,121 @@ def __exit__(self, *_: Any) -> None:
+
+        tmp_dir = get_tmp_dir()
+
+        new_driver_info = subprocess.check_output(
+            [exe, "--internal-print-update-driver-info-and-exit"],
+            text=True,
+        )
+        (
+            start_scripts,
+            vlans_str,
+            testscript,
+            output_directory,
+        ) = new_driver_info.rstrip().split("\n")
```
I think it would be better to write these arguments in JSON in the interactive driver derivation output instead. This way we don't have less stuff to run outside of the sandbox, and less guessing about the binary path and its arguments.

You could even add a separate derivation that only outputs this JSON to avoid having to fetch new versions of host packages (qemu) that we won't use here anyway.

Radvendii replied:
The reason I did it the way I did is that things seem to be built so that the driver itself is a standalone executable. It gets wrapped in a wrapper that sets arguments. So I tried with this PR to maintain the separation between those layers. But they could also get set by command line arguments or environment variables.

What you're proposing puts us in a state where the executable depends on the wrapper being set up in a particular way, which feels like mixing layers to me.

Maybe I'm being overly concerned about this though, and the layers are already mixed up or we just don't care if we mix them up. If so, I think your solution is indeed better, but I didn't know enough to make that judgment call myself.

Radvendii commented on nixos/doc/manual/development/running-nixos-tests-interactively.section.md:
````diff
@@ -90,6 +107,35 @@ $ ./result/bin/nixos-test-driver --keep-vm-state
+```ShellSession
+$ build_cmd="nix-build . -A nixosTests.login.driverInteractive"
+$ eval $build_cmd
+$ ./result/bin/nixos-test-driver --rebuild-cmd $build_cmd 2>machine_output
````
Good call on the quotes.

What stderr output are you worried about? Python errors, for instance, are output to stdout.

I added it to the docs because in my experience the interactive driver is extremely irritating to use with the VM output clogging up the interactive terminal, and it took me a while to realise I could redirect it.

Radvendii replied:
Oh indeed calls to `logger.info()` and the like get sent to stderr. I'll just remove this, since it's not actually related to the PR.

Radvendii committed:
better logging

Radvendii committed:
forgot quotes in docs

Radvendii committed:
log actual rebuild command

Radvendii committed:
remove separating outputs section

there are caveats, and it's not actually related to this PR

Radvendii committed:
need exe as an explicit argument

Radvendii committed:
correctly modify interactive repl's environment

Radvendii committed:
correctly detect if machine is booted

Radvendii committed:
reasonable default for --rebuild-exe

Radvendii commented:
Realised we can set `rebuildExe` from the nixos test driver

Radvendii commented:
@roberth can you take a look at my responses to your review requests?

roberth committed:
nixosTests.nixos-test-driver.interactive-redeploy: init

roberth commented on nixos/lib/test-driver/test_driver/__init__.py:
```diff
@@ -112,6 +134,15 @@ def main() -> None:
     args = arg_parser.parse_args()
 
+    # print the info needed to update another Driver
+    # TODO: properly escape these arguments
```
Is this TODO important?

Radvendii replied:
The only cases I can think of where this would cause a problem are
1. if someone had a truly psychotic CWD with a newline in it.
2. if they're doing something very weird with the test driver by passing arguments in manually **and** they're using paths with spaces (or newlines) in them.

Technically not impossible, but I would be surprised if it came up in practice. And escaping stuff means decoding it on the other end. It just adds a bunch of complexity so I didn't bother.

roberth replied:
I guess it works then, yeah.

If you could send a JSON instead of a custom stream, that could also simplify the reading code, and make those parts of the program easier to correlate (i.e. when browsing with `grep` or similar)

roberth commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -101,6 +110,153 @@ def __exit__(self, *_: Any) -> None:
+                                machine.start_command = start_cmd
+                else:
+                    self.logger.error(
+                        f"Skipping the update of {start_cmd.machine_name}, because there are mltiple machines with that name. This shouldn't be possible, and is an error in the testing framework."
```
```suggestion
                        f"Skipping the update of {start_cmd.machine_name}, because there are multiple machines with that name. This shouldn't be possible, and is an error in the testing framework."
```

roberth reviewed (commented):
I've pushed a basic tests; seems to work fine so far. Maybe you could test a few more things that you ran into while working on this?

Other than that just minor comments. Really glad you were able to change it to this approach; much cleaner!

Radvendii committed:
typo

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

Radvendii committed:
nixfmt & correct typo

Radvendii committed:
add release notes

Radvendii committed:
add more testing for interactive-redeploy

Radvendii committed:
reformat

Radvendii commented:
@roberth Wow, thanks for that test! At some point I tried to work out how to automate testing for this feature and kind of gave up as "tests within tests are broken". Good to see that it is possible.
I added my own stuff to the test for more edge-cases that I ran into. I think it should be good to go now. In theory JSON is better, but I think better to get this work in and do it as a follow-up if needed.

Radvendii commented:
@roberth can we push this through?

tfc commented:
Hey @Radvendii, i am rejoining this thread very late. Just tried it out on my machine and got an error where i am not sure if i did anything wrong or if it is a bug.

What i did so far was the following:

1. run `nix-build . -A pkgs.nixosTests.bittorrent.driverInteractive` (in the nixpkgs repo of course)
2. run `./result/bin/nixos-test-driver --rebuild-cmd "nix-build . -A pkgs.nixosTests.bittorrent.driverInteractive"
3. after `start_all()` and waiting a bit, i changed the configuration of one machine (`client1` in that case) by adding some new package to its `environment.systemPackages`.
4. Run `rebuild()`.
5. This leads to an error `/nix/store/[...]-nixos-vm/bin/switch-to-configuration: No such file or directory`

Can you give me a pointer what i might have done wrong here?

Radvendii commented:
You're holding it completely right. This is definitely a bug I introduced in later commits. Thanks for finding it! I know vaguely where it is but I haven't had the time and brainspace to fix it and work out why the test isn't failing. I'll ping you when I do.
